api/src/main/java/org/openmrs/obs/handler/AbstractHandler.java

```diff
@@ -67,37 +70,47 @@ public File getOutputFileToWrite(Obs obs) throws IOException {
 		
 		File dir = OpenmrsUtil.getDirectoryInApplicationDataDirectory(Context.getAdministrationService().getGlobalProperty(
 		    OpenmrsConstants.GLOBAL_PROPERTY_COMPLEX_OBS_DIR));
-		File outputfile = null;
+		File outputFile;
 		
 		// Get the output stream
 		if (null == title) {
 			String now = longfmt.format(new Date());
-			outputfile = new File(dir, now);
+			outputFile = new File(dir, now);
 		} else {
 			title = title.replace("." + extension, "");
-			outputfile = new File(dir, title + "." + extension);
-			// outputfile = new File(dir, title);
+			outputFile = new File(dir, title + "." + extension);
 		}
 		
-		int i = 0;
-		String tmp = null;
+		String tmp;
 		
-		// If the Obs does not exist, but the File does, append a two-digit
-		// count number to the filename and save it.
-		while (obs.getObsId() == null && outputfile.exists() && i < 100) {
-			tmp = null;
+		if (outputFile.exists() || StringUtils.isNotBlank(obs.getComplexData().getMimeType())) {
 			// Remove the extension from the filename.
-			tmp = String.valueOf(outputfile.getAbsolutePath().replace("." + extension, ""));
-			outputfile = null;
-			// Append two-digit count number to the filename.
-			String filename = (i < 1) ? tmp + "_" + nf.format(Integer.valueOf(++i)) : tmp.replace(nf.format(Integer
-			        .valueOf(i)), nf.format(Integer.valueOf(++i)));
+			tmp = String.valueOf(outputFile.getAbsolutePath().replace("." + extension, ""));
+			
+			// Get MIME type
+			String mimetype = "";
+			if (StringUtils.isNotBlank(obs.getComplexData().getMimeType())) {
+				mimetype = "_" + obs.getComplexData().getMimeType().replace("/", "-slash-");
+			}
+			
+			// If the Obs does not exist, but the File does, append a two-digit
+			// count number to the filename and save it.
+			String twoDigitNumber = "";
+			if (new File(tmp + mimetype + "." + extension).exists()) {
+				for (int i = 1; obs.getObsId() == null && i < 100; i++) {
```

Comment:
@tmarzeion am a bit confused by this change because it seems to do more than i would expect. For instance, instance, can you throw some more light as to why you removed the while loop?